### PR TITLE
Implement "getTemplateData"

### PR DESCRIPTION
**Pull request by yuchi**
See #178 for reference.

### CODE REVIEW

File: packages/metal-soy/src/Soy.js [hunk 1]
```diff
@@ -40,6 +40,8 @@ class Soy extends IncrementalDomRenderer.constructor {
 	 * template call's data. The copying needs to be done because, if the component
 	 * itself is passed directly, some problems occur when soy tries to merge it
 	 * with other data, due to property getters and setters. This is safer.
+	 * Also calls the component's "getTemplateData" is called to let it change the
```
Comment (mairatma) [inline packages/metal-soy/src/Soy.js]: The wording here is a bit weird, maybe just remove teh `is called` part? :)

**Review by mairatma — COMMENTED** [review]
This pull request is very good, it's a very small change that can already be of much help. Thanks a lot for working on this :)

I just have a single inline comment, but @eduardolundgren, can you take a look as well and say what you think?

File: packages/metal-soy/src/Soy.js (hunk 1 above)
Comment (yuchi) [reply packages/metal-soy/src/Soy.js]: Sorry, wrote it while listening to the training so I probably wrote it in two phases…

Gosh I made a conflict merge artifact without using git…

**Comment by yuchi** [comment]
Thank you @mairatma, but please have a look at the discussion in #178 before accepting this.

File: packages/metal-soy/src/Soy.js (hunk 1 above)
Comment (yuchi) [reply packages/metal-soy/src/Soy.js]: Should be fixed now

File: packages/metal-soy/src/Soy.js [hunk 2]
```diff
@@ -59,6 +61,11 @@ class Soy extends IncrementalDomRenderer.constructor {
 				data[params[i]] = component[params[i]].bind(component);
 			}
 		}
+
+		if (component.getTemplateData) {
```
Comment (yuchi) [inline packages/metal-soy/src/Soy.js]: @mairatma should `isFn` be used here? Is usually in Metal the presence of a property defined by the contract enough?

Comment (mairatma) [reply packages/metal-soy/src/Soy.js]: `isFunction` would indeed be safer. We usually just check for the presence, but I think it'd be best to use `isFunction` everywhere actually.

**Comment by eduardolundgren** [comment]
LGTM. This feature is nice to have, just keep in mind that when template data is mutated on the client side it would require to reflect the server render state, which is okay.

For the method name I kinda liked the `willRender` but since we have on Component `setState`, `getState`, maybe this might reference the name `State`? Few extra options:

* `beforeRender(data)`
* `formatState(data)`
* `mutateState(data)`
* `prepareState(data)`
* `prepareStateForRender(data)`
* `willRender(data)`

wdyt?

**Comment by yuchi** [comment]
I find the `prepare*` prefix way more apt. I vote for `prepareStateForRender`.

**Comment by brunobasto** [comment]
I like prepareSate better I think.

**Comment by brunobasto** [comment]
On a second thought, maybe formatState makes more sense.

**Comment by eduardolundgren** [comment]
Also, is this just for Soy? Would make sense to have it on the lifecycle for JSX as well?

**Comment by yuchi** [comment]
@eduardolundgren please have a look at #178 where this topics are discussed. It is exclusively for Soy because in other contexts it would make no sense, or even worse it would be a bad practice.

**Comment by eduardolundgren** [comment]
@yuchi I've missed that paragraph on your comment. Make sense. Now I guess it's only the naming left, then we can merge.

File: packages/metal-soy/src/Soy.js (hunk 2 above)
Comment (eduardolundgren) [reply packages/metal-soy/src/Soy.js]: Yep, `isFunction` is needed here.

Comment (yuchi) [reply packages/metal-soy/src/Soy.js]: I’ll fix this as a reminder, I’m gonna close this PR

**Comment by yuchi** [comment]
I’m gonna close this, waiting for the end of the poll in #178.

Don‘t forget to cast your vote!

**Pull request closed without merging**